File: solutions/python/causal_inference_ex2.py
# =============================================================================
# Chapter 17 - Exercise 2: Propensity score matching
# Beta-blocker use and 1-year mortality
# =============================================================================

# Libraries -------------------------------------------------------------------
# pip install numpy pandas scikit-learn statsmodels matplotlib
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.neighbors import NearestNeighbors
import statsmodels.api as sm
import statsmodels.formula.api as smf


def expit(x):
    return 1 / (1 + np.exp(-x))


# --- The dataset from the exercise ------------------------------------------
# Note that all covariates are CENTRED in the linear predictors, so the
# intercepts set the baseline rate for an average patient rather than for a
# 0-year-old. Without that, this cohort has a 92% one-year mortality.
rng = np.random.default_rng(123)
n = 1500

age = rng.normal(70, 8, n)
creatinine = rng.normal(1.2, 0.4, n)
heart_failure = rng.binomial(1, 0.35, n)
prior_mi = rng.binomial(1, 0.20, n)

treatment = rng.binomial(1, expit(-0.4 + 0.05 * (age - 70)
                                 + 0.7 * heart_failure
                                 + 0.9 * prior_mi
                                 + 0.8 * (creatinine - 1.2)))
lp_untreated = (-1.9 + 0.05 * (age - 70) + 0.7 * heart_failure
                + 0.8 * prior_mi + 1.0 * (creatinine - 1.2))
death_1yr = rng.binomial(1, expit(lp_untreated - 0.8 * treatment))

df = pd.DataFrame(dict(age=age, creatinine=creatinine,
                       heart_failure=heart_failure, prior_mi=prior_mi,
                       treatment=treatment, death_1yr=death_1yr))

# The truth, available only because we simulated the data.
treated_mask = df.treatment == 1
p1 = expit(lp_untreated[treated_mask] - 0.8).mean()
p0 = expit(lp_untreated[treated_mask]).mean()
TRUE_ATT_RD = p1 - p0
TRUE_ATT_OR = (p1 / (1 - p1)) / (p0 / (1 - p0))

print(f"Cohort: {n} patients | {100 * df.treatment.mean():.0f}% treated "
      f"| {100 * df.death_1yr.mean():.1f}% died within 1 year")
print(f"TRUE ATT: risk difference {TRUE_ATT_RD:+.4f}, "
      f"odds ratio {TRUE_ATT_OR:.3f}\n")

covs = ["age", "creatinine", "heart_failure", "prior_mi"]

# =============================================================================
# (a) Estimate the propensity score and look at OVERLAP
# =============================================================================
ps_model = smf.logit("treatment ~ age + creatinine + heart_failure + prior_mi",
                     data=df).fit(disp=0)
df["ps"] = ps_model.predict(df)
df["logit_ps"] = np.log(df.ps / (1 - df.ps))

print("--- (a) Propensity score distribution ---")
print(df.groupby("treatment")["ps"].agg(["count", "min", "median", "max"]).round(3))

fig, ax = plt.subplots(figsize=(8, 4.5))
for value, label, colour in [(0, "No beta-blocker", "#0072B2"),
                             (1, "Beta-blocker", "#D55E00")]:
    df.loc[df.treatment == value, "ps"].plot.density(
        ax=ax, alpha=0.6, label=label, color=colour)
ax.set_xlabel("Propensity score")
ax.set_ylabel("Density")
ax.set_title("Propensity score overlap by treatment group")
ax.legend()
plt.tight_layout()
plt.show()

print("\nBoth groups span roughly the same range of scores, with no pile-up at")
print("0 or 1, so positivity is not obviously violated and matching is feasible.")

# =============================================================================
# (b) 1:1 nearest-neighbour matching with a caliper of 0.2 SD
# =============================================================================
# Greedy 1:1 nearest-neighbour matching without replacement, with a caliper of
# 0.2 SD of the propensity score (the same convention MatchIt uses in R).
# Treated patients are processed in DESCENDING propensity-score order, so the
# hardest-to-match patients get first pick of the controls.
caliper = 0.2 * df.ps.std()

treated_idx = df.index[df.treatment == 1].to_numpy()
treated_idx = treated_idx[np.argsort(-df.loc[treated_idx, "ps"].to_numpy())]
control_idx = df.index[df.treatment == 0].to_numpy()

nn = NearestNeighbors(n_neighbors=len(control_idx)).fit(
    df.loc[control_idx, ["ps"]].to_numpy())
dist, order = nn.kneighbors(df.loc[treated_idx, ["ps"]].to_numpy())

used = set()
pairs = []
for row, t_i in enumerate(treated_idx):
    for d, c_pos in zip(dist[row], order[row]):
        if d > caliper:
            break                       # nothing else is close enough either
        c_i = control_idx[c_pos]
        if c_i not in used:
            used.add(c_i)
            pairs.append((t_i, c_i))
            break

matched_idx = [i for pair in pairs for i in pair]
m_data = df.loc[matched_idx].copy()

n_treated = len(treated_idx)
n_matched = len(pairs)
print(f"\n--- (b) Matching ---")
print(f"{n_matched} of {n_treated} treated patients found a partner; "
      f"{n_treated - n_matched} did NOT.")
print("""Those unmatched patients are silently DROPPED. They are not a random
subset -- they are the ones with the most extreme propensity scores, i.e. the
patients who were most obviously going to be treated. So the estimate no longer
describes 'all treated patients'; it describes the treated patients for whom a
comparable untreated patient exists. Always report how many were dropped, and
compare their characteristics.""")

# =============================================================================
# (c) Balance before and after (the Love plot)
# =============================================================================
def smd(data, var):
    t = data.loc[data.treatment == 1, var]
    c = data.loc[data.treatment == 0, var]
    pooled_sd = np.sqrt((t.var() + c.var()) / 2)
    return (t.mean() - c.mean()) / pooled_sd


balance = pd.DataFrame({
    "before": [smd(df, v) for v in covs],
    "after": [smd(m_data, v) for v in covs],
}, index=covs)

print("\n--- (c) Standardised mean differences ---")
print(balance.round(3))

fig, ax = plt.subplots(figsize=(7, 3.6))
y = np.arange(len(covs))
ax.scatter(balance["before"].abs(), y, label="Before matching",
           color="#D55E00", s=60)
ax.scatter(balance["after"].abs(), y, label="After matching",
           color="#0072B2", s=60)
ax.axvline(0.1, linestyle="--", color="grey")
ax.set_yticks(y)
ax.set_yticklabels(covs)
ax.set_xlabel("|Standardised mean difference|")
ax.set_title("Covariate balance: before and after matching")
ax.legend()
plt.tight_layout()
plt.show()

worst = balance["after"].abs().idxmax()
worst_value = balance.loc[worst, "after"]
print(f"\nThe worst-balanced covariate after matching is {worst} at "
      f"{abs(worst_value):.3f}")
if abs(worst_value) > 0.1:
    print("-- marginally OUTSIDE the conventional 0.1 threshold. That is a real")
    print("(if mild) failure, and greedy 1:1 matching often ends up here when a")
    print("lot of treated patients go unmatched. The fixes, in order: allow more")
    print("controls per treated patient, match on the Mahalanobis distance rather")
    print("than the propensity score alone, or abandon matching for weighting")
    print("(Exercise 3), which uses every patient and usually balances better.")
else:
    print("-- inside the conventional 0.1 threshold, so the matched groups have a")
    print("comparable mix of patients.")

# =============================================================================
# (d) The ATT, as an odds ratio and as a risk difference
# =============================================================================
or_fit = smf.glm("death_1yr ~ treatment", data=m_data,
                 family=sm.families.Binomial()).fit()
log_or = or_fit.params["treatment"]
or_ci = or_fit.conf_int().loc["treatment"]

rd_fit = smf.ols("death_1yr ~ treatment", data=m_data).fit()
rd = rd_fit.params["treatment"]
rd_ci = rd_fit.conf_int().loc["treatment"]

unadjusted_rd = (df.loc[df.treatment == 1, "death_1yr"].mean()
                 - df.loc[df.treatment == 0, "death_1yr"].mean())

# The estimate describes the MATCHED treated patients, not all treated patients,
# so recompute the truth over exactly that subgroup for a fair comparison.
matched_treated = m_data.index[m_data.treatment == 1]
lp_m = lp_untreated[df.index.get_indexer(matched_treated)]
p1_m = expit(lp_m - 0.8).mean()
p0_m = expit(lp_m).mean()
TRUE_MATCHED_RD = p1_m - p0_m
TRUE_MATCHED_OR = (p1_m / (1 - p1_m)) / (p0_m / (1 - p0_m))

print("\n--- (d) ATT estimates in the matched sample ---")
print(f"Odds ratio      : {np.exp(log_or):.3f} "
      f"(95% CI {np.exp(or_ci[0]):.3f}, {np.exp(or_ci[1]):.3f})"
      f"   [truth {TRUE_ATT_OR:.3f}]")
print(f"Risk difference : {rd:+.4f} "
      f"(95% CI {rd_ci[0]:+.4f}, {rd_ci[1]:+.4f})"
      f"   [truth {TRUE_ATT_RD:+.4f}]")
print(f"\nUnadjusted risk difference in the full cohort: {unadjusted_rd:+.4f}")
print("-- a fraction of the true effect. Matching recovers most of what the")
print("naive comparison hides.")

print(f"""
But compare against the right target. The truth quoted above is the ATT over
ALL treated patients. Our estimate only describes the {len(pairs)} treated patients
who found a match, and for THAT subgroup the true values are:
  risk difference {TRUE_MATCHED_RD:+.4f}   odds ratio {TRUE_MATCHED_OR:.3f}
which is what the estimate should be judged against. The gap between the two
targets is the price of discarding unmatched patients: matching answers a
slightly different question from the one you asked.""")

print("\nMortality in the matched sample:")
print(f"  Treated: {m_data.loc[m_data.treatment == 1, 'death_1yr'].mean():.3f}"
      f"    Control: {m_data.loc[m_data.treatment == 0, 'death_1yr'].mean():.3f}")

# =============================================================================
# (e) E-value: how strong would a hidden confounder have to be?
# =============================================================================
# The E-value works on the risk-ratio scale and is symmetric about the null, so
# a protective estimate is first flipped to the above-null side.
def e_value(rr):
    if rr < 1:
        rr = 1 / rr
    return rr + np.sqrt(rr * (rr - 1))


# With a ~15% outcome the odds ratio overstates the risk ratio, so compute the
# risk ratio directly rather than reusing the OR.
risk_t = m_data.loc[m_data.treatment == 1, "death_1yr"].mean()
risk_c = m_data.loc[m_data.treatment == 0, "death_1yr"].mean()
rr = risk_t / risk_c

a = m_data.loc[m_data.treatment == 1, "death_1yr"].sum()
b = (m_data.treatment == 1).sum()
c = m_data.loc[m_data.treatment == 0, "death_1yr"].sum()
d = (m_data.treatment == 0).sum()
se_log_rr = np.sqrt(1 / a - 1 / b + 1 / c - 1 / d)
rr_lo = np.exp(np.log(rr) - 1.96 * se_log_rr)
rr_hi = np.exp(np.log(rr) + 1.96 * se_log_rr)

e_point = e_value(rr)
e_ci = 1.0 if rr_hi >= 1 else e_value(rr_hi)

print("\n--- (e) E-value ---")
print(f"Risk ratio: {rr:.3f} (95% CI {rr_lo:.3f}, {rr_hi:.3f})")
print(f"E-value for the point estimate       : {e_point:.2f}")
print(f"E-value for the CI bound nearest null: {e_ci:.2f}")
print(f"""
Interpretation in one sentence: an unmeasured confounder would have to be
associated with BOTH beta-blocker use and death by a risk ratio of at least
{e_point:.2f} -- over and above age, creatinine, heart failure and prior MI -- to
explain away this result entirely.

Whether that is plausible is a clinical judgement, not a statistical one.
Compare it with the strength of the confounders you DID measure: if none of them
reaches that magnitude, a hidden one probably does not either.""")

File: solutions/R/causal_inference_ex2.R
# =============================================================================
# Chapter 17 - Exercise 2: Propensity score matching
# Beta-blocker use and 1-year mortality
# =============================================================================
#
# Libraries -------------------------------------------------------------------
# Every one of these is needed. In particular, WITHOUT library(broom) the call
# to tidy() below fails with the misleading
#     Error in UseMethod("tidy") : no applicable method for 'tidy' ...
library(tidyverse) # tibble(), mutate(), ggplot2
library(MatchIt)   # matchit(), match.data()
library(cobalt)    # love.plot(), bal.tab()
library(broom)     # tidy()

# --- The dataset from the exercise ------------------------------------------
set.seed(123)
n <- 1500

exercise_dat <- tibble(
  age           = rnorm(n, 70, 8),
  creatinine    = rnorm(n, 1.2, 0.4),
  heart_failure = rbinom(n, 1, 0.35),
  prior_mi      = rbinom(n, 1, 0.20)
) |>
  mutate(
    treatment = rbinom(n, 1, plogis(-0.4 + 0.05 * (age - 70) +
                                      0.7 * heart_failure +
                                      0.9 * prior_mi +
                                      0.8 * (creatinine - 1.2))),
    death_1yr = rbinom(n, 1, plogis(-1.9 + 0.05 * (age - 70) +
                                      0.7 * heart_failure +
                                      0.8 * prior_mi +
                                      1.0 * (creatinine - 1.2) -
                                      0.8 * treatment))
  )

# The truth, available only because we simulated the data. Compute it once so
# every estimate below can be judged against it.
lp_untreated <- with(exercise_dat, -1.9 + 0.05 * (age - 70) +
  0.7 * heart_failure + 0.8 * prior_mi + 1.0 * (creatinine - 1.2))
treated <- exercise_dat$treatment == 1
p1 <- mean(plogis(lp_untreated[treated] - 0.8))
p0 <- mean(plogis(lp_untreated[treated]))
TRUE_ATT_RD <- p1 - p0
TRUE_ATT_OR <- (p1 / (1 - p1)) / (p0 / (1 - p0))

cat("Cohort:", nrow(exercise_dat), "patients |",
    sprintf("%.0f%% treated | %.1f%% died within 1 year\n",
            100 * mean(exercise_dat$treatment),
            100 * mean(exercise_dat$death_1yr)))
cat(sprintf("TRUE ATT: risk difference %+.4f, odds ratio %.3f\n\n",
            TRUE_ATT_RD, TRUE_ATT_OR))

# =============================================================================
# (a) Estimate the propensity score and look at OVERLAP
# =============================================================================
ps_model <- glm(treatment ~ age + creatinine + heart_failure + prior_mi,
  data = exercise_dat, family = binomial
)
exercise_dat$ps <- predict(ps_model, type = "response")

cat("--- (a) Propensity score distribution ---\n")
print(exercise_dat |>
  group_by(treatment) |>
  summarise(
    n = n(),
    min = round(min(ps), 3),
    median = round(median(ps), 3),
    max = round(max(ps), 3)
  ))

# Overlap plot: we want the two densities to cover the same range of scores.
ps_plot <- ggplot(exercise_dat, aes(
  x = ps,
  fill = factor(treatment, labels = c("No beta-blocker", "Beta-blocker"))
)) +
  geom_density(alpha = 0.5) +
  scale_fill_manual(values = c("#0072B2", "#D55E00")) +
  labs(
    x = "Propensity score", y = "Density", fill = NULL,
    title = "Propensity score overlap by treatment group"
  ) +
  theme_minimal()
print(ps_plot)

cat("\nBoth groups span roughly the same range of scores, with no pile-up at\n")
cat("0 or 1, so positivity is not obviously violated and matching is feasible.\n")

# =============================================================================
# (b) 1:1 nearest-neighbour matching with a caliper of 0.2 SD
# =============================================================================
m_out <- matchit(treatment ~ age + creatinine + heart_failure + prior_mi,
  data = exercise_dat,
  method = "nearest",
  distance = "glm", # propensity score from logistic regression
  caliper = 0.2,    # no match further than 0.2 SD of logit(PS)
  ratio = 1
)

cat("\n--- (b) Matching ---\n")
print(m_out)

n_treated <- sum(exercise_dat$treatment == 1)
n_matched_treated <- sum(match.data(m_out)$treatment == 1)
cat(sprintf(
  "\n%d of %d treated patients found a partner; %d did NOT.\n",
  n_matched_treated, n_treated, n_treated - n_matched_treated
))
cat("Those unmatched patients are silently DROPPED. They are not a random\n")
cat("subset -- they are the ones with the most extreme propensity scores, i.e.\n")
cat("the patients who were most obviously going to be treated. So the estimate\n")
cat("no longer describes 'all treated patients'; it describes the treated\n")
cat("patients for whom a comparable untreated patient exists. Always report\n")
cat("how many were dropped, and compare their characteristics.\n")

# =============================================================================
# (c) Love plot: did matching actually balance the covariates?
# =============================================================================
love_p <- love.plot(m_out,
  thresholds = c(m = 0.1),
  binary = "std",
  var.order = "unadjusted",
  title = "Covariate balance: before and after matching",
  colors = c("#D55E00", "#0072B2")
)
print(love_p)

cat("\n--- (c) Balance table ---\n")
print(bal.tab(m_out, thresholds = c(m = 0.1)))
cat("\nAll four COVARIATES are now inside the 0.1 threshold (age is the worst at\n")
cat("0.075, down from 0.38 before matching), so the matched groups have a\n")
cat("comparable mix of patients. Note that the row labelled 'distance' -- the\n")
cat("propensity score itself -- is still around 0.12. That is common and is not\n")
cat("in itself a failure: the score is a summary, and it is balance on the\n")
cat("actual covariates that removes confounding. If a real COVARIATE were above\n")
cat("0.1, the fixes are to tighten the caliper, allow more controls per treated\n")
cat("patient, or move to weighting (Exercise 3) rather than matching.\n")

# =============================================================================
# (d) The ATT, as an odds ratio and as a risk difference
# =============================================================================
m_data <- match.data(m_out)

or_fit <- glm(death_1yr ~ treatment,
  data = m_data, family = binomial, weights = weights
)
or_res <- tidy(or_fit, conf.int = TRUE, exponentiate = TRUE) |>
  filter(term == "treatment")

# A risk difference is easier to communicate to clinicians than an odds ratio.
rd_fit <- lm(death_1yr ~ treatment, data = m_data, weights = weights)
rd_res <- tidy(rd_fit, conf.int = TRUE) |> filter(term == "treatment")

cat("\n--- (d) ATT estimates in the matched sample ---\n")
cat(sprintf(
  "Odds ratio      : %.3f (95%% CI %.3f, %.3f)   [truth %.3f]\n",
  or_res$estimate, or_res$conf.low, or_res$conf.high, TRUE_ATT_OR
))
cat(sprintf(
  "Risk difference : %+.4f (95%% CI %+.4f, %+.4f)   [truth %+.4f]\n",
  rd_res$estimate, rd_res$conf.low, rd_res$conf.high, TRUE_ATT_RD
))
cat(sprintf(
  "\nFor comparison, the UNADJUSTED risk difference in the full cohort is %+.4f\n",
  mean(exercise_dat$death_1yr[exercise_dat$treatment == 1]) -
    mean(exercise_dat$death_1yr[exercise_dat$treatment == 0])
))
cat("-- less than half the true effect. Matching recovers most of what the\n")
cat("naive comparison hides.\n")

# But compare against the RIGHT target. The estimate describes the matched
# treated patients, not all treated patients, so recompute the truth over
# exactly that subgroup.
matched_rows <- as.integer(rownames(m_data)[m_data$treatment == 1])
lp_m <- lp_untreated[matched_rows]
p1_m <- mean(plogis(lp_m - 0.8))
p0_m <- mean(plogis(lp_m))
cat(sprintf(
  "\nThe truth quoted above is the ATT over ALL %d treated patients. Our\n",
  n_treated
))
cat(sprintf(
  "estimate only describes the %d who found a match, and for THAT subgroup the\n",
  n_matched_treated
))
cat(sprintf(
  "true values are: risk difference %+.4f, odds ratio %.3f.\n",
  p1_m - p0_m,
  ((p1_m / (1 - p1_m)) / (p0_m / (1 - p0_m)))
))
cat("That is what the estimate should be judged against, and the gap between\n")
cat("the two targets is the price of discarding unmatched patients: matching\n")
cat("answers a slightly different question from the one you asked.\n")

cat("\nMortality in the matched sample:\n")
cat(sprintf(
  "  Treated: %.3f    Control: %.3f\n",
  mean(m_data$death_1yr[m_data$treatment == 1]),
  mean(m_data$death_1yr[m_data$treatment == 0])
))

# =============================================================================
# (e) E-value: how strong would a hidden confounder have to be?
# =============================================================================
# The E-value works on the risk-ratio scale and is symmetric about the null, so
# a protective estimate is first flipped to the above-null side.
e_value <- function(rr) {
  if (rr < 1) rr <- 1 / rr
  rr + sqrt(rr * (rr - 1))
}

# With a 15% outcome the odds ratio overstates the risk ratio, so compute the
# risk ratio directly from the matched sample rather than reusing the OR.
risk_treated <- mean(m_data$death_1yr[m_data$treatment == 1])
risk_control <- mean(m_data$death_1yr[m_data$treatment == 0])
rr_point <- risk_treated / risk_control

# CI for the risk ratio, on the log scale
a <- sum(m_data$death_1yr[m_data$treatment == 1])
b <- sum(m_data$treatment == 1)
c_ <- sum(m_data$death_1yr[m_data$treatment == 0])
d_ <- sum(m_data$treatment == 0)
se_log_rr <- sqrt(1 / a - 1 / b + 1 / c_ - 1 / d_)
rr_lo <- exp(log(rr_point) - 1.96 * se_log_rr)
rr_hi <- exp(log(rr_point) + 1.96 * se_log_rr)

# For a protective effect, the CI bound CLOSEST to the null is the upper one.
e_point <- e_value(rr_point)
e_ci <- if (rr_hi >= 1) 1 else e_value(rr_hi)

cat("\n--- (e) E-value ---\n")
cat(sprintf(
  "Risk ratio: %.3f (95%% CI %.3f, %.3f)\n", rr_point, rr_lo, rr_hi
))
cat(sprintf("E-value for the point estimate      : %.2f\n", e_point))
cat(sprintf("E-value for the CI bound nearest null: %.2f\n", e_ci))

cat("\nInterpretation in one sentence: an unmeasured confounder would have to\n")
cat(sprintf(
  "be associated with BOTH beta-blocker use and death by a risk ratio of at\nleast %.2f",
  e_point
))
cat(" -- over and above age, creatinine, heart failure and prior MI --\n")
cat("to explain away this result entirely.\n")
cat("\nWhether that is plausible is a clinical judgement, not a statistical\n")
cat("one. Compare it with the strength of the confounders you DID measure: if\n")
cat("none of them reaches that magnitude, a hidden one probably does not either.\n")

# The EValue package does this for you, including for odds ratios and
# hazard ratios, and is worth using in real work:
#   install.packages("EValue")
#   library(EValue)
#   evalues.RR(est = 1 / rr_point, lo = 1 / rr_hi, hi = 1 / rr_lo)
